### PR TITLE
feat(middleware): Alternative middleware equivalent to Helmet

### PR DESCRIPTION
**Pull request by watany-dev**
I created this middleware as an experimental equivalent to Helmet. Originally made for my own use, I've opened a PR in this repository because there was a request for Helmet in the Github Issues. One difference from the official Helmet is that this middleware only adopts headers that can be set with fixed values, and users can only set each header to true/false.

I'm still undecided about the name of the middleware. It's not a pure Helmet. Names like 'mini-helmet' or 'kabuto' could be interesting, but for now, I've opted for a straightforward naming.

### Author should do the followings, if applicable

- [x] Add tests
- [x] Run tests
- [x] `yarn denoify` to generate files for Deno

## Doc(sample)

---
### Secure Header Middleware
This middleware simplifies the setup of security headers. Inspired in part by the capabilities of Helmet, it allows you to control the activation and deactivation of specific security headers.

#### Installation
The middleware is built into the Hono framework, so no additional installation is required.

#### Usage
```ts
import { secureHeader } from 'hono/secure-header';

app.use('*', secureHeader({ xFrameOptions: false, xXssProtection: false }));

```

fix: https://github.com/honojs/hono/issues/798

### CODE REVIEW

**Comment by yusukebe** [comment]
Cooooool! Really wanted it. I'll check it later!

**Comment by watany-dev** [comment]
Please also refer to the documentation.
https://github.com/honojs/website/pull/83

File: src/middleware/secure-header/index.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    .filter(([key]) => options[key as keyof SecureHeaderOptions])
+    .map(([, value]) => value)
+
+  return async (ctx, next) => {
+    headersToSet.forEach(([header, value]) => {
+      ctx.res.headers.set(header, value)
+    })
+
+    ctx.res.headers.delete('X-Powered-By')
+    await next()
```
Comment (yusukebe) [inline src/middleware/secure-header/index.ts:61]: The procedures to manipulate headers should be placed on after `await next()`. For example, this test fails:

```ts
it('should remove x-powered-by header', async () => {
  const app = new Hono()
  app.use('*', secureHeader())
  app.use('*', poweredBy())

  const res = await app.request('/')
  expect(res.headers.get('x-powered-by')).toBeFalsy()
})
```

File: src/middleware/secure-header/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+describe('Secure Header Middleware', () => {
+  it('all headers enabled', async () => {
+    const app = new Hono()
+    app.use('*', secureHeader())
+    app.get('/test', async (ctx) => {
+      return ctx.text('test')
+    })
+
+    const req = new Request('http://localhost/test', { method: 'GET' })
+    const res = await app.request(req)
```
Comment (yusukebe) [inline src/middleware/secure-header/index.test.ts:13]: You can write:

```ts
const req = new Request('http://localhost/test', { method: 'GET' })
const res = await app.request(req)
```

to, as one line:

```ts
const res = await app.request('/test')
```

**Comment by yusukebe** [comment]
Hi @watany-dev,

Thank you for the excellent PR! It's evident that this is a much-desired feature, and I'm grateful that you've taken it on.

About the naming: I understand the familiarity of "Helmet" for developers who have experience with Express. However, for those who might not have that background (like myself), it could be a tad confusing.

I concur that `secure-header` is more descriptive and self-explanatory. That being said, considering the plural nature (as we're dealing with multiple headers), `secure-headers` might be more apt. For instance, the `Response` object utilizes the `headers` property, not `header`:

```ts
const res = new Response()
res.headers.get('foo')
```

I've also added some comments in the PR; kindly see them.

Lastly, it's imperative for users to be informed about the positioning of this middleware. As illustrated by the following test, it's evident that order matters:

```ts
it('should remove x-powered-by header', async () => {
  const app = new Hono()
  app.use('*', poweredBy())
  app.use('*', secureHeader())

  const res = await app.request('/test')
  expect(res.headers.get('x-powered-by')).toBeFalsy()
})
```

We'll need to highlight this aspect in our documentation on the website.

**Comment by watany-dev** [comment]
@yusukebe 
Thank you for the comment. I've made the corrections; could you please check?

File: src/middleware/secure-header/index.test.ts (hunk 2 above)
Comment (watany-dev) [reply src/middleware/secure-header/index.test.ts]: ok! refactored.

File: src/middleware/secure-header/index.ts (hunk 1 above)
Comment (watany-dev) [reply src/middleware/secure-header/index.ts]: fixed.

**Comment by zisra** [comment]
Some questions/concerns
- Are there plans for including support for the CSP header? (Which is also probably the most important)...
- ... as well as `X-Powered-by` and `Cross-Origin-Embedder-Policy` seem to be missing
- Can we add links to the MDN docs of the respective to the docs?
-`Origin-Agent-Cluster` is included within the docs AND "not supported" list 🤔

I would be very happy to work on it if you don't want to, just let me know

**Comment by yusukebe** [comment]
Hi @watany-dev,

Thank you for the revisions. Also, please take into account the feedback provided by @zisra.

I believe:
> "Can we add links to the MDN docs of the respective to the docs?"

is a great idea.

**Review by yusukebe — APPROVED** [review]

**Comment by watany-dev** [comment]
@zisra 
Thank you for the feedback! I believe we've been able to set up the CSP header for general use cases. We've also addressed the other comments. (We'll provide the CSP documentation at a later date.)

**Comment by watany-dev** [comment]
- Address CI errors
- Instructions on using CSP
- Document differences in default values between our implementation and Helmet
- Document the order of middleware usage

File: src/middleware/secure-headers/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import { Hono } from '../../hono'
```
Comment (yusukebe) [inline src/middleware/secure-headers/index.test.ts:1]: You can add the follow line on the top to fix lint errors:

```ts
/* eslint-disable quotes */
```

This allow using double quotes.

Comment (watany-dev) [reply src/middleware/secure-headers/index.test.ts:2]: thank you !! fix it.

**Comment by watany-dev** [comment]
@yusukebe 
Thank you. I believe I've corrected the areas that needed fixing. There is a part where CI fails with a 429 error, but it seems to be a temporary error unrelated to this matter. I will add "workflow_dispatch" to ci.yaml in a separate PR so that it can be executed manually.

**Comment by watany-dev** [comment]
This is the latest version, including the document.

**Review by yusukebe — CHANGES_REQUESTED** [review]
@watany-dev 

As you said, please do `denoify`! This is a last thing to do!

**Comment by watany-dev** [comment]
Sorry for the wait. It's done now.

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@watany-dev 

Thank you very much. You've done the awesome work. The next release (minor change) includes this feature will be released later. Thanks again!

**Pull request merged**